### PR TITLE
added collective coverage Python script to work with counter BBid

### PR DESCRIPTION
**Pull request by nikign**
Collective coverage script runs the inputs in an input folder on the program and saves the coverage for each in an output folder. It also saves the collective coverage of all inputs in a file specified by the user.

### CODE REVIEW

File: collective_coverage.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
+import inspect
+import getopt
+import sys
+import os
+
+
+def get_full_path(filename):
+    current_file_path = os.path.dirname(os.path.abspath(inspect.getfile(inspect.currentframe())))
```
Comment (rsprabery) [inline collective_coverage.py]: With python 3, i think `pathlib` and `__file__` is the preferred way of doing this. See [this](https://stackoverflow.com/questions/3430372/how-do-i-get-the-full-path-of-the-current-files-directory) stackoverflow question.

File: collective_coverage.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+    bb_coverage_path = get_full_path('./counter_BBid_coverage')
+    test_inputs = os.listdir(input_dir)
+    object_file_name = object_file.strip().split('/')[-1].split('.')[0]
+
+    for input_name in test_inputs:
+        input_file = os.path.join(input_dir, input_name)
+        input_name = input_name.strip().split('.')[0]
+        if not os.path.isfile(input_file):
+            continue
```
Comment (rsprabery) [inline collective_coverage.py]: Let's log an error here since this shouldn't be the normal case.

File: collective_coverage.py [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+        coverage_file = os.path.join(output_dir, "%s_%s_coverage.txt" % (object_file_name, input_name))
+        os.system('%s -o %s  -A %s -- %s @@' %(bb_coverage_path, coverage_file, input_file, object_file))
+
+
+def calculate_collective_coverage(output_dir, collective_coverage_path):
+    coverage_map = {}
+    seed_count = 0
+    for coverage_file_name in os.listdir(output_dir):
+        coverage_file_path = os.path.join(output_dir, coverage_file_name)
+        if not os.path.isfile(coverage_file_path):
```
Comment (rsprabery) [inline collective_coverage.py:57]: Same as above, let's log an error here as this shouldn't be the case under normal operation.

File: collective_coverage.py [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+            continue
+        seed_count += 1
+        with open(coverage_file_path) as coverage_file:
+            for line in coverage_file:
+                coverage_list = line.strip().split(":")
+                if len(coverage_list) != 2:
+                    print('Invalid output file format in %s. Format should be "block id: coverage count" in each line.')
```
Comment (rsprabery) [inline collective_coverage.py:64]: 👍

**Review by rsprabery — CHANGES_REQUESTED** [review]
Just a few minor nits, but LGTM overall!

File: collective_coverage.py (hunk 2 above)
Comment (nikign) [reply collective_coverage.py]: Should we just log that something is wrong or should we exit the run?

Comment (rsprabery) [reply collective_coverage.py]: I think existing makes sense unless you've observed some behavior that makes this a more common case.

File: collective_coverage.py [hunk 5]
```diff
@@ -1,11 +1,13 @@
...
 import getopt
+import pathlib
 import sys
 import os
 
 
 def get_full_path(filename):
-    current_file_path = os.path.dirname(os.path.abspath(inspect.getfile(inspect.currentframe())))
+    current_file_path = pathlib.Path(__file__).parent.absolute()
+    # current_file_path = os.path.dirname(os.path.abspath(inspect.getfile(inspect.currentframe())))
```
Comment (rsprabery) [inline collective_coverage.py]: please remove the commented line.

Comment (nikign) [reply collective_coverage.py]: Sorry, I had missed it!

File: llvm_mode/afl-llvm-pass.so.cc [hunk 6]
```diff
@@ -116,14 +116,24 @@ bool AFLCoverage::runOnModule(Module &M) {
...
 
+      MDNode* BBid_meta;
+      for (Instruction& instr : BB.getInstList()) {
+            assert(instr.hasMetadata() && "The first instruction of the block should include block id, have you instrumented the code with set-counter-BBid-llvm-pass first?");
+            BBid_meta = instr.getMetadata("BBid");
+            break;
+      }
+      char c;
+      std::string meta_string = cast<MDString>(BBid_meta->getOperand(0))->getString();
+      std::sscanf(meta_string.c_str(), "%" SCNd64 "%c", &block_counter, &c);
```
Comment (laurentsimon) [inline llvm_mode/afl-llvm-pass.so.cc]: SCNd64 is for signed integer. I there an equivalent unsigned?

Comment (laurentsimon) [inline llvm_mode/afl-llvm-pass.so.cc]: sscanf does not detect overflow. In C, we should use [strtoull](https://linux.die.net/man/3/strtoull) with the right checks. Here's an [example](https://github.com/Samsung/afl_cc/blob/master/afl-fuzz.c#L1959), which I encourage you to check if it's correct...

File: llvm_mode/afl-llvm-pass.so.cc [hunk 7]
```diff
@@ -116,14 +116,24 @@ bool AFLCoverage::runOnModule(Module &M) {
...
 
-      assert(block_counter < MAP_SIZE && "counter is too large");
+    for (auto &BB : F) {
       BasicBlock::iterator IP = BB.getFirstInsertionPt();
       IRBuilder<> IRB(&(*IP));
 
       /* Make up cur_loc based on block counter */
 
+      MDNode* BBid_meta;
+      for (Instruction& instr : BB.getInstList()) {
```
Comment (laurentsimon) [inline llvm_mode/afl-llvm-pass.so.cc]: let's move this to a dedicated function `getBBCounter()`?

File: llvm_mode/set-counter-BBid-llvm-pass.so.cc [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+  /* Instrument all the things! */
+
+  u64 block_counter = 1;
+  
+  for (auto &F : M){
+    for (auto &BB : F) {
+
+        assert(block_counter < MAP_SIZE && "counter is too large");
+        std::string value = std::to_string(block_counter);
+        MDString * S = MDString::get(C, value.c_str());
```
Comment (laurentsimon) [inline llvm_mode/set-counter-BBid-llvm-pass.so.cc:51]: add some `assert(S)` and `assert(Meta)`. Did you check the assert() is triggered? assert() is a noop if NDEBUG is set. You shuold check if this is the case or not.

File: llvm_mode/set-counter-BBid-llvm-pass.so.cc [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+        MDNode* Meta = MDNode::get(C, S);
+        for (Instruction& instr : BB.getInstList()) {
```
Comment (laurentsimon) [inline llvm_mode/set-counter-BBid-llvm-pass.so.cc:59]: let's add an assertion that the BB always has at least one instruction, i.e. that the loop is always entered.

File: llvm_mode/afl-llvm-pass.so.cc (hunk 6 above)
Comment (leonbett) [reply llvm_mode/afl-llvm-pass.so.cc]: There is also mdconst::dyn_extract\<ConstantInt\>(md->getOperand(0)). Not sure if it works in this case though.

File: llvm_mode/set-counter-BBid-llvm-pass.so.cc (hunk 8 above)
Comment (nikign) [reply llvm_mode/set-counter-BBid-llvm-pass.so.cc:51]: It is triggered, but it doesn't have the best UI to show the message! I can check manually and use FATAL instead if that sounds better.

**Review by rsprabery — APPROVED** [review]

**Pull request merged**